src/actions/heap/heap.ts

```diff
@@ -216,38 +249,45 @@ export class HeapAction extends Hub.Action {
     return form
   }
 
-  private validateRequest(request: Hub.ActionRequest): Error | undefined {
-    if (!request.formParams.env_id || request.formParams.env_id.match(/\D/g)) {
+  /*
+  * validate the connection configuration, it's an input from the looker admin.
+  * The validation will be run when a field is received (for backward compatibility)
+  * The same validation will be when the connection is initially setup.
+  */
+  private validateParams(formParams: Hub.ParamMap): Error | undefined {
+    if (!formParams.env_id || formParams.env_id.match(/\D/g)) {
       return new Error(
-        `Heap environment ID is invalid: ${request.formParams.env_id}`,
+        `Heap environment ID is invalid: ${formParams.env_id}`,
       )
     }
 
     if (
-      !request.formParams.property_type ||
+      !formParams.property_type ||
       !(this.supportedPropertyTypes as string[]).includes(
-        request.formParams.property_type,
+        formParams.property_type,
       )
     ) {
       return new Error(
-        `Unsupported property type: ${request.formParams.property_type}`,
+        `Unsupported property type: ${formParams.property_type}`,
       )
     }
 
     if (
-      !request.formParams.heap_field ||
-      request.formParams.heap_field.length === 0
+      !formParams.heap_field ||
+      formParams.heap_field.length === 0
     ) {
       return new Error("Column mapping to a Heap field must be provided.")
     }
   }
 
-  private extractHeapFieldByLabel(fields: Hub.Field[], heapFieldLabel: string): Hub.Field {
+  private extractHeapFieldByLabel(fields: Hub.Field[], heapFieldLabel: string, errors: Error[]): Hub.Field | undefined {
```

Comment:
instead of passing in an errors array, could we instead update the calling code to catch this error when it is thrown and take the appropriate action there?

Reply:
it's part of `onFields`, a callback when the stream is processed. I think what I could do is that, skip onRow when `heapFieldLabel` is empty.

Reply:
the `errors` will be bubbled up to the final part that generates the response.